packages/aws-cdk-lib/aws-cloudfront-origins/lib/s3-origin.ts

```diff
@@ -44,46 +108,210 @@ export class S3Origin implements cloudfront.IOrigin {
 /**
  * An Origin specific to a S3 bucket (not configured for website hosting).
  *
- * Contains additional logic around bucket permissions and origin access identities.
+ * Contains additional logic around bucket permissions and origin access control (via OAI or OAC).
  */
-class S3BucketOrigin extends cloudfront.OriginBase {
-  private originAccessIdentity!: cloudfront.IOriginAccessIdentity;
+abstract class S3BucketOrigin extends cloudfront.OriginBase {
+  public static withAccessIdentity(bucket: s3.IBucket, props: S3OriginProps = {}): S3BucketOrigin {
+    return new (class OriginAccessIdentity extends S3BucketOrigin {
```

Comment:
This nested class looks weird to me. Can you instead split them out?
```
abstract class S3BucketOrigin extends cloudfront.OriginBase {
   ...
}

class OriginAccessIdentity extends S3BucketOrigin {
   ...
}

class OriginAccessControl extends S3BucketOrigin {
  ...
}
```
And on line 90~100, you can do 
```
    else if (props.originAccessControl) {
      this.origin = new OriginAccessControl(bucket, props);
    } else if (props.originAccessIdentity) {
      this.origin = new OriginAccessIdentity(bucket, props);
    } else {
      this.origin = FeatureFlags.of(bucket.stack)
        .isEnabled(cxapi.CLOUDFRONT_USE_ORIGIN_ACCESS_CONTROL_BY_DEFAULT) ?
        OriginAccessControl(bucket, props) :
        new OriginAccessIdentity(bucket, props);
    }
```

Reply:
This was one of the suggested changes from the RFC (see [feedback](https://github.com/aws/aws-cdk-rfcs/pull/619#discussion_r1638661763)). I agree the structure looks a bit unconventional, but what I like is that the API is very clear on lines 90-100: 
```
else if (props.originAccessControl) {
      this.origin = S3BucketOrigin.withAccessControl(bucket, props);
    } else if (props.originAccessIdentity) {
      this.origin = S3BucketOrigin.withAccessIdentity(bucket, props);
    } else {
      this.origin = FeatureFlags.of(bucket.stack)
        .isEnabled(cxapi.CLOUDFRONT_USE_ORIGIN_ACCESS_CONTROL_BY_DEFAULT) ?
        S3BucketOrigin.withAccessControl(bucket, props) :
        S3BucketOrigin.withAccessIdentity(bucket, props);
    }
```
since it is a S3 origin with either OAI or OAC configured